In [1]:
from pathlib import Path

import pandas as pd

In [2]:
pd.set_option("display.max_columns", None)
pd.set_option("display.max_colwidth", 100)
pd.set_option("display.width", 1200)

In [3]:
RAW_DATA_DIR = Path("../data/raw")

RAW_DATA_DIR.exists()

True

In [4]:
csv_files = sorted(RAW_DATA_DIR.glob("*.csv"))

print(f"Number of CSV files found: {len(csv_files)}")

for file_path in csv_files:
    print(file_path.name)

Number of CSV files found: 9
olist_customers_dataset.csv
olist_geolocation_dataset.csv
olist_order_items_dataset.csv
olist_order_payments_dataset.csv
olist_order_reviews_dataset.csv
olist_orders_dataset.csv
olist_products_dataset.csv
olist_sellers_dataset.csv
product_category_name_translation.csv


In [5]:
file_inventory = []

for file_path in csv_files:
    file_inventory.append(
        {
            "file_name": file_path.name,
            "file_size_mb": round(file_path.stat().st_size / (1024 ** 2), 2),
        }
    )

file_inventory_df = pd.DataFrame(file_inventory)

file_inventory_df

,file_name,file_size_mb
0,olist_customers_dataset.csv,8.62
1,olist_geolocation_dataset.csv,58.44
2,olist_order_items_dataset.csv,14.72
3,olist_order_payments_dataset.csv,5.51
4,olist_order_reviews_dataset.csv,13.78
5,olist_orders_dataset.csv,16.84
6,olist_products_dataset.csv,2.27
7,olist_sellers_dataset.csv,0.17
8,product_category_name_translation.csv,0.00


In [6]:
dataset_files = {
    "customers": RAW_DATA_DIR / "olist_customers_dataset.csv",
    "geolocation": RAW_DATA_DIR / "olist_geolocation_dataset.csv",
    "order_items": RAW_DATA_DIR / "olist_order_items_dataset.csv",
    "payments": RAW_DATA_DIR / "olist_order_payments_dataset.csv",
    "reviews": RAW_DATA_DIR / "olist_order_reviews_dataset.csv",
    "orders": RAW_DATA_DIR / "olist_orders_dataset.csv",
    "products": RAW_DATA_DIR / "olist_products_dataset.csv",
    "sellers": RAW_DATA_DIR / "olist_sellers_dataset.csv",
    "category_translation": RAW_DATA_DIR / "product_category_name_translation.csv",
}

dataset_files

{'customers': WindowsPath('../data/raw/olist_customers_dataset.csv'),
 'geolocation': WindowsPath('../data/raw/olist_geolocation_dataset.csv'),
 'order_items': WindowsPath('../data/raw/olist_order_items_dataset.csv'),
 'payments': WindowsPath('../data/raw/olist_order_payments_dataset.csv'),
 'reviews': WindowsPath('../data/raw/olist_order_reviews_dataset.csv'),
 'orders': WindowsPath('../data/raw/olist_orders_dataset.csv'),
 'products': WindowsPath('../data/raw/olist_products_dataset.csv'),
 'sellers': WindowsPath('../data/raw/olist_sellers_dataset.csv'),
 'category_translation': WindowsPath('../data/raw/product_category_name_translation.csv')}

In [7]:
for dataset_name, file_path in dataset_files.items():
    print(
        f"{dataset_name:<22} "
        f"exists={file_path.exists()} "
        f"file={file_path.name}"
    )

customers              exists=True file=olist_customers_dataset.csv
geolocation            exists=True file=olist_geolocation_dataset.csv
order_items            exists=True file=olist_order_items_dataset.csv
payments               exists=True file=olist_order_payments_dataset.csv
reviews                exists=True file=olist_order_reviews_dataset.csv
orders                 exists=True file=olist_orders_dataset.csv
products               exists=True file=olist_products_dataset.csv
sellers                exists=True file=olist_sellers_dataset.csv
category_translation   exists=True file=product_category_name_translation.csv


## Looking at samples of each dataset

In [8]:
sample_data = {}

for dataset_name, file_path in dataset_files.items():
    try:
        sample_df = pd.read_csv(file_path, nrows=5)
        sample_data[dataset_name] = sample_df

        print(
            f"{dataset_name:<22} "
            f"loaded successfully — "
            f"{sample_df.shape[0]} rows, "
            f"{sample_df.shape[1]} columns"
        )

    except Exception as error:
        print(f"{dataset_name:<22} failed: {error}")

customers              loaded successfully — 5 rows, 5 columns
geolocation            loaded successfully — 5 rows, 5 columns
order_items            loaded successfully — 5 rows, 7 columns
payments               loaded successfully — 5 rows, 5 columns
reviews                loaded successfully — 5 rows, 7 columns
orders                 loaded successfully — 5 rows, 8 columns
products               loaded successfully — 5 rows, 9 columns
sellers                loaded successfully — 5 rows, 4 columns
category_translation   loaded successfully — 5 rows, 2 columns


In [9]:
sample_data["customers"]

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [10]:
for dataset_name, sample_df in sample_data.items():
    print(f"\n{dataset_name.upper()}")
    print("-" * 60)

    for column in sample_df.columns:
        print(column)


CUSTOMERS
------------------------------------------------------------
customer_id
customer_unique_id
customer_zip_code_prefix
customer_city
customer_state

GEOLOCATION
------------------------------------------------------------
geolocation_zip_code_prefix
geolocation_lat
geolocation_lng
geolocation_city
geolocation_state

ORDER_ITEMS
------------------------------------------------------------
order_id
order_item_id
product_id
seller_id
shipping_limit_date
price
freight_value

PAYMENTS
------------------------------------------------------------
order_id
payment_sequential
payment_type
payment_installments
payment_value

REVIEWS
------------------------------------------------------------
review_id
order_id
review_score
review_comment_title
review_comment_message
review_creation_date
review_answer_timestamp

ORDERS
------------------------------------------------------------
order_id
customer_id
order_status
order_purchase_timestamp
order_approved_at
order_delivered_carrier_date
ord

## basic schema summary

In [11]:
schema_summary = []

for dataset_name, file_path in dataset_files.items():
    sample_df = pd.read_csv(file_path, nrows=100)

    for column_name, dtype in sample_df.dtypes.items():
        schema_summary.append(
            {
                "dataset": dataset_name,
                "column_name": column_name,
                "sample_inferred_dtype": str(dtype),
            }
        )

schema_summary_df = pd.DataFrame(schema_summary)

schema_summary_df

,dataset,column_name,sample_inferred_dtype
0,customers,customer_id,str
1,customers,customer_unique_id,str
2,customers,customer_zip_code_prefix,int64
3,customers,customer_city,str
4,customers,customer_state,str
5,geolocation,geolocation_zip_code_prefix,int64
6,geolocation,geolocation_lat,float64
7,geolocation,geolocation_lng,float64
8,geolocation,geolocation_city,str
9,geolocation,geolocation_state,str


# Load complete datasets in dictionary of dataframes

In [ ]:
# Loading all datasets in a dictionary of data frames
datasets = {}

for dataset_name, file_path in dataset_files.items():
    try:
        df = pd.read_csv(file_path)

        datasets[dataset_name] = df

        print(
            f"{dataset_name:<22} "
            f"{df.shape[0]:>10,} rows × "
            f"{df.shape[1]:>2} columns"
        )

    except Exception as error:
        print(f"{dataset_name:<22} failed: {error}")

customers                  99,441 rows ×  5 columns
geolocation             1,000,163 rows ×  5 columns
order_items               112,650 rows ×  7 columns
payments                  103,886 rows ×  5 columns
reviews                    99,224 rows ×  7 columns
orders                     99,441 rows ×  8 columns
products                   32,951 rows ×  9 columns
sellers                     3,095 rows ×  4 columns
category_translation           71 rows ×  2 columns


In [14]:
datasets.keys()

dict_keys(['customers', 'geolocation', 'order_items', 'payments', 'reviews', 'orders', 'products', 'sellers', 'category_translation'])

In [15]:
datasets["orders"].head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


## General inspection of all datasets

This section performs an initial inspection of every source dataset, including:

- Dataset dimensions
- Column names
- Data types
- Missing values
- Exact duplicate rows
- Number of unique values
- Summary statistics


### Inspect shape and column names

# Customers Dataset

In [19]:
customers = datasets["customers"]
customers.head()


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [20]:
customers.shape

(99441, 5)

Inspecting : 
data types  
non-null counts  
memory usage  


In [21]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.8 MB


#### Statistics

In [22]:
# summary statistics
customers.describe()

,customer_zip_code_prefix
count,99441.000000
mean,35137.474583
std,29797.938996
min,1003.000000
25%,11347.000000
50%,24416.000000
75%,58900.000000
max,99990.000000


In [24]:
customers.describe(include="str")

,customer_id,customer_unique_id,customer_city,customer_state
count,99441,99441,99441,99441
unique,99441,96096,4119,27
top,06b8999e2fba1a1fbc88172c00ba8bc7,8d50f5eadf50201ccdcedfb9e2ac8455,sao paulo,SP
freq,1,17,15540,41746


#### Missing Values

In [25]:
customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

#### Duplicate rows

In [29]:
customers.duplicated().sum()

np.int64(0)

#### Unique values

In [26]:
customers.nunique()

customer_id                 99441
customer_unique_id          96096
customer_zip_code_prefix    14994
customer_city                4119
customer_state                 27
dtype: int64

Key conclusions:

- customer_id has 99,441 unique values, matching the row count, so it is the likely primary key.
- customer_unique_id has 96,096 unique values, so some real customers placed multiple orders.
- There are no missing values.
- customer_state has 27 values, which is reasonable for Brazil’s states plus the Federal District.
- customer_zip_code_prefix should eventually be treated as an identifier/string rather than a numeric measurement.

# Orders Dataset

In [39]:
orders = datasets["orders"]
orders.head(2)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00


In [31]:
orders.shape

(99441, 8)

In [32]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB


In [34]:
orders.describe(include="str")

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
count,99441,99441,99441,99441,99281,97658,96476,99441
unique,99441,99441,8,98875,90733,81018,95664,459
top,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2018-03-31 15:08:21,2018-02-27 04:31:10,2018-05-09 15:48:00,2018-05-14 20:02:44,2017-12-20 00:00:00
freq,1,1,96478,3,9,47,3,522


In [35]:
# checking null
orders.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [36]:
# duplicates
orders.duplicated().sum()

np.int64(0)

In [37]:
# unique
orders.nunique()

order_id                         99441
customer_id                      99441
order_status                         8
order_purchase_timestamp         98875
order_approved_at                90733
order_delivered_carrier_date     81018
order_delivered_customer_date    95664
order_estimated_delivery_date      459
dtype: int64

In [38]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

    Key conclusions: 
- The table has 99,441 rows and 8 columns.
- order_id has 99,441 unique values, so it is the likely primary key.
- customer_id is also unique here because each row represents one order and each order has one order-specific customer ID.
- There are no exact duplicate rows.
- All date columns are currently stored as strings and should later be converted to datetime.
- Missing values exist in:
- order_approved_at: 160
- order_delivered_carrier_date: 1,783
- order_delivered_customer_date: 2,965
- These missing dates may be valid for cancelled, unavailable, created, processing, or not-yet-delivered orders, so we should not remove them blindly.
- order_estimated_delivery_date has only 459 unique values because many orders share the same estimated delivery date. That is normal.
- The table grain is: One row represents one order.

# Order Items Dataset

In [55]:
order_items = datasets["order_items"]

order_items.head(2)

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.9,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.9,19.93


In [41]:
order_items.shape

(112650, 7)

In [42]:
order_items.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  str    
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  str    
 3   seller_id            112650 non-null  str    
 4   shipping_limit_date  112650 non-null  str    
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), str(4)
memory usage: 6.0 MB


In [43]:
order_items.describe()

,order_item_id,price,freight_value
count,112650.000000,112650.000000,112650.000000
mean,1.197834,120.653739,19.990320
std,0.705124,183.633928,15.806405
min,1.000000,0.850000,0.000000
25%,1.000000,39.900000,13.080000
50%,1.000000,74.990000,16.260000
75%,1.000000,134.900000,21.150000
max,21.000000,6735.000000,409.680000


In [44]:
order_items.describe(include="str")

,order_id,product_id,seller_id,shipping_limit_date
count,112650,112650,112650,112650
unique,98666,32951,3095,93318
top,8272b63d03f5f79c56e9e4120aec44ef,aca2eb7d00ea1a7b8ebd4e68314663af,6560211a19b47992c3666cc44a7e94c0,2018-03-01 02:50:48
freq,21,527,2033,21


In [45]:
# missing  values
order_items.isnull().sum()

order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [46]:
# duplicates
order_items.duplicated().sum()

np.int64(0)

In [47]:
# unique values
order_items.nunique()

order_id               98666
order_item_id             21
product_id             32951
seller_id               3095
shipping_limit_date    93318
price                   5968
freight_value           6999
dtype: int64

In [ ]:
#Check repeated orders
order_items["order_id"].value_counts().head(10)

order_id
8272b63d03f5f79c56e9e4120aec44ef    21
1b15974a0141d54e36626dca3fdc731a    20
ab14fdcfbe524636d65ee38360e22ce8    20
428a2f660dc84138d969ccd69a0ab6d5    15
9ef13efd6949e4573a18964dd1bbe7f5    15
73c8ab38f07dc94389065f7eba4f297a    14
9bdc4d4c71aa1de4606060929dee888c    14
37ee401157a3a0b28c9c6d0ed8c3b24b    13
2c2a19b5703863c908512d135aa6accc    12
3a213fcdfe7d98be74ea0dc05a8b31ae    12
Name: count, dtype: int64

In [49]:
#Check item sequence values
order_items["order_item_id"].value_counts().sort_index()

order_item_id
1     98666
2      9803
3      2287
4       965
5       460
6       256
7        58
8        36
9        28
10       25
11       17
12       13
13        8
14        7
15        5
16        3
17        3
18        3
19        3
20        3
21        1
Name: count, dtype: int64

In [50]:
order_items["order_id"].is_unique

False

In [51]:
order_items.duplicated(
    subset=["order_id", "order_item_id"]
).sum()

np.int64(0)

In [ ]:
#Inspect price and freight ranges
order_items[["price", "freight_value"]].describe()

,price,freight_value
count,112650.000000,112650.000000
mean,120.653739,19.990320
std,183.633928,15.806405
min,0.850000,0.000000
25%,39.900000,13.080000
50%,74.990000,16.260000
75%,134.900000,21.150000
max,6735.000000,409.680000


In [53]:
order_items.nlargest(
    10,
    "price"
)[["order_id", "product_id", "seller_id", "price", "freight_value"]]

,order_id,product_id,seller_id,price,freight_value
3556,0812eb902a67711a1cb742b3cdaa65ae,489ae2aa008f021502940f251d4cce7f,e3b4998c7a498169dc7bce44e6bb6277,6735.00,194.31
112233,fefacc66af859508bf1a7934eab1e97f,69c590f7ffc7bf8db97190b6cb6ed62e,80ceebb4ee9b31afb6c6a916a574a1e2,6729.00,193.21
107841,f5136e38d1a14a4dbd87dff67da82701,1bdf5e6731585cf01aa8169c7028d6ad,ee27a8f15b1dded4d213a468ba4eb391,6499.00,227.66
74336,a96610ab360d42a2e5335a3998b4718a,a6492cc69376c469ab6f61d8f44de961,59417c56835dd8e2e72f91f809cd4092,4799.00,151.34
11249,199af31afc78c699f0dbf71fb178d4d4,c3ed642d592594bb648ff4a04cee2747,59417c56835dd8e2e72f91f809cd4092,4690.00,74.34
62086,8dbc85d1447242f3b127dda390d56e19,259037a6a41845e455183f89c5035f18,c72de06d72748d1a0dfb2125be43ba63,4590.00,91.78
29193,426a9742b533fc6fed17d1fd6d143d7e,a1beef8f3992dbd4cd8726796aa69c53,512d298ac2a96d1931b6bd30aa21f61d,4399.87,113.45
45843,68101694e5c5dc7330c91e1bbc36214f,6cdf8fc1d741c76586d8b6b15e9eef30,ed4acab38528488b65a9a9c603ff024a,4099.99,75.27
78310,b239ca7cd485940b31882363b52e6674,dd113cb02b2af9c8e5787e8f1f0722f6,821fb029fc6e495ca4f08a35d51e53a5,4059.00,104.51
59137,86c4eab1571921a6a6e248ed312f5a5a,6902c1962dd19d540807d0ab8fade5c6,fa1c13f2614d7b5c4749cbc52fecda94,3999.90,17.01


In [54]:
order_items.nlargest(
    10,
    "freight_value"
)[["order_id", "product_id", "seller_id", "price", "freight_value"]]

,order_id,product_id,seller_id,price,freight_value
73486,a77e1550db865202c56b19ddc6dc4d53,ec31d2a17b299511e7c8627be9337b9b,257e61d3251fb5efb9daadddbc2cf7ca,979.00,409.68
3303,076d1555fb53a89b0ef4d529e527a0f6,a3cd9517ebf5a50dca25acce54f3b171,6fa9202c10491e472dffd59a3e82b2a3,2338.08,375.28
28044,3fde74c28a3d5d618c00f26d51baafa0,a3cd9517ebf5a50dca25acce54f3b171,6fa9202c10491e472dffd59a3e82b2a3,2338.08,375.28
69797,9f49bd16053df810384e793386312674,256a9c364b75753b97bee410c9491ad8,5c030029b5916fed0986310385ec9009,1149.00,339.59
16731,264a7e199467906c0727394df82d1a6a,97c948ebc8c04b26b7bbb095d4228f2a,17f51e7198701186712e53a39c564617,1050.00,338.30
87936,c7a07ddd52bbe18b61da49a8d89853d3,97c948ebc8c04b26b7bbb095d4228f2a,17f51e7198701186712e53a39c564617,1050.00,322.10
5037,0b6230647ed16f4b3e70282dc4b5b87f,46e24ce614899e36617e37ea1e4aa6ff,17f51e7198701186712e53a39c564617,1050.00,321.88
3584,0822bcde10bb5d023755a71bc8f7797f,363a9f5b97bf194da23858be722a7aa5,9596c870880d900012f2e8e6e30d06d7,990.00,321.46
29787,43bdbd9dc0931d72befdf4765af6c442,7e53e051875b2a0c9f22acd8a9a29a20,eeb6de78f79159600292e314a77cbd18,3089.00,317.47
48320,6ddfbf514959b49b6410c01ad93054bb,363a9f5b97bf194da23858be722a7aa5,9596c870880d900012f2e8e6e30d06d7,1045.00,314.40


#### Conclusions
- The table contains 112,650 rows and 7 columns.
- One row represents one item within an order.
- There are no missing values and no exact duplicate rows.
- `order_id` is not unique because one order can contain multiple items.
- The combination of `order_id` and `order_item_id` is unique, so it is the likely composite primary key.
- `order_item_id` represents the item sequence inside an order and ranges from 1 to 21.
- There are 98,666 unique orders, 32,951 unique products, and 3,095 unique sellers.
- `shipping_limit_date` is currently stored as text and should later be converted to `datetime`.
- `price` ranges from 0.85 to 6,735.00, while `freight_value` ranges from 0 to 409.68.
- Some very high prices and freight values exist, but they should be investigated as possible valid outliers rather than removed automatically.

# Products datset

In [56]:
products = datasets["products"]

In [57]:
products.head(2)

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0


In [58]:
products.shape

(32951, 9)

In [59]:
products.info()


<class 'pandas.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  str    
 1   product_category_name       32341 non-null  str    
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), str(2)
memory usage: 2.3 MB


In [60]:
products.describe()

,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
count,32341.000000,32341.000000,32341.000000,32949.000000,32949.000000,32949.000000,32949.000000
mean,48.476949,771.495285,2.188986,2276.472488,30.815078,16.937661,23.196728
std,10.245741,635.115225,1.736766,4282.038731,16.914458,13.637554,12.079047
min,5.000000,4.000000,1.000000,0.000000,7.000000,2.000000,6.000000
25%,42.000000,339.000000,1.000000,300.000000,18.000000,8.000000,15.000000
50%,51.000000,595.000000,1.000000,700.000000,25.000000,13.000000,20.000000
75%,57.000000,972.000000,3.000000,1900.000000,38.000000,21.000000,30.000000
max,76.000000,3992.000000,20.000000,40425.000000,105.000000,105.000000,118.000000


In [61]:
products.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [66]:
products[products.product_category_name.isnull() == True]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
105,a41e356c76fab66334f36de622ecbd3a,NaN,NaN,NaN,NaN,650.0,17.0,14.0,12.0
128,d8dee61c2034d6d075997acef1870e9b,NaN,NaN,NaN,NaN,300.0,16.0,7.0,20.0
145,56139431d72cd51f19eb9f7dae4d1617,NaN,NaN,NaN,NaN,200.0,20.0,20.0,20.0
154,46b48281eb6d663ced748f324108c733,NaN,NaN,NaN,NaN,18500.0,41.0,30.0,41.0
197,5fb61f482620cb672f5e586bb132eae9,NaN,NaN,NaN,NaN,300.0,35.0,7.0,12.0
...,...,...,...,...,...,...,...,...,...
32515,b0a0c5dd78e644373b199380612c350a,NaN,NaN,NaN,NaN,1800.0,30.0,20.0,70.0
32589,10dbe0fbaa2c505123c17fdc34a63c56,NaN,NaN,NaN,NaN,800.0,30.0,10.0,23.0
32616,bd2ada37b58ae94cc838b9c0569fecd8,NaN,NaN,NaN,NaN,200.0,21.0,8.0,16.0
32772,fa51e914046aab32764c41356b9d4ea4,NaN,NaN,NaN,NaN,1300.0,45.0,16.0,45.0


In [68]:
# duplicates
products.duplicated().sum()

np.int64(0)

In [69]:
# unique items
products.nunique()

product_id                    32951
product_category_name            73
product_name_lenght              66
product_description_lenght     2960
product_photos_qty               19
product_weight_g               2204
product_length_cm                99
product_height_cm               102
product_width_cm                 95
dtype: int64

#### Conclusions

- The table contains 32,951 rows and 9 columns.
- Each row represents one unique product.
- `product_id` is unique and can be treated as the primary key.
- There are no duplicate rows.
- Most product attributes are numeric measurements (weight and dimensions).
- 610 products have missing values for category, name length, description length, and photo count.
- Only 2 products have missing values in weight and dimension columns.
- Product dimensions and weight appear reasonable overall, although a few large values should be inspected later as potential outliers.
- There are 73 product categories.
- Since products are referenced by orders, missing product information should not be removed without careful consideration during data cleaning.

# Payments Dataset

In [70]:
payments = datasets["payments"]
payments.head(2)

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39


In [71]:
payments.shape

(103886, 5)

In [72]:
payments.info()

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 4.0 MB


In [73]:
payments.describe()

,payment_sequential,payment_installments,payment_value
count,103886.000000,103886.000000,103886.000000
mean,1.092679,2.853349,154.100380
std,0.706584,2.687051,217.494064
min,1.000000,0.000000,0.000000
25%,1.000000,1.000000,56.790000
50%,1.000000,1.000000,100.000000
75%,1.000000,4.000000,171.837500
max,29.000000,24.000000,13664.080000


In [74]:
payments.describe(include="str")

,order_id,payment_type
count,103886,103886
unique,99440,5
top,fa65dad1b0e818e3ccc5cb0e39231352,credit_card
freq,29,76795


In [75]:
payments.isnull().sum()

order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

In [77]:
payments.duplicated().sum()

np.int64(0)

In [78]:
payments.nunique()

order_id                99440
payment_sequential         29
payment_type                5
payment_installments       24
payment_value           29077
dtype: int64

In [ ]:
# payment methods used,
payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

In [84]:
# installment distribution
payments["payment_installments"].value_counts().sort_index()

payment_installments
0         2
1     52546
2     12413
3     10461
4      7098
5      5239
6      3920
7      1626
8      4268
9       644
10     5328
11       23
12      133
13       16
14       15
15       74
16        5
17        8
18       27
20       17
21        3
22        1
23        1
24       18
Name: count, dtype: int64

In [83]:
# revenew analysis
payments["payment_value"].describe()

count    103886.000000
mean        154.100380
std         217.494064
min           0.000000
25%          56.790000
50%         100.000000
75%         171.837500
max       13664.080000
Name: payment_value, dtype: float64

In [ ]:
# To check unusually large payments.
payments.nlargest(
    10,
    "payment_value"
)[["order_id", "payment_type", "payment_installments", "payment_value"]]

,order_id,payment_type,payment_installments,payment_value
52107,03caa2c082116e1d31e67e9ae3700499,credit_card,1,13664.08
34370,736e1922ae60d0d6a89247b851902527,boleto,1,7274.88
41419,0812eb902a67711a1cb742b3cdaa65ae,credit_card,8,6929.31
49581,fefacc66af859508bf1a7934eab1e97f,boleto,1,6922.21
85539,f5136e38d1a14a4dbd87dff67da82701,boleto,1,6726.66
62409,2cc9089445046817a7539d90805e6e5a,boleto,1,6081.54
43232,a96610ab360d42a2e5335a3998b4718a,credit_card,10,4950.34
70320,b4c4b76c642808cbe472a32b86cddc95,credit_card,5,4809.44
6440,199af31afc78c699f0dbf71fb178d4d4,credit_card,8,4764.34
67546,8dbc85d1447242f3b127dda390d56e19,credit_card,8,4681.78


#### Key analysis
- The table contains 103,886 rows and 5 columns.
- One row represents one payment transaction or payment sequence associated with an order.
- There are no missing values and no exact duplicate rows.
- `order_id` is not unique because an order can have multiple payment records.
- The likely composite key is `order_id` together with `payment_sequential`.
- There are 99,440 unique orders, so almost every order has a payment record.
- There are 5 payment types: `credit_card`, `boleto`, `voucher`, `debit_card`, and `not_defined`.
- Credit card is the most common payment method by a large margin.
- `payment_installments` ranges from 0 to 24; the 2 records with 0 installments should be investigated during cleaning.
- `payment_value` ranges from 0 to 13,664.08.
- Very large payment values exist, but they should be treated as possible valid outliers and investigated rather than removed automatically.
- Some orders use multiple payment sequences, with the maximum `payment_sequential` equal to 29.

# Reviews Dataset

In [85]:
reviews = datasets["reviews"]
reviews.head(2)

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13


In [86]:
reviews.shape

(99224, 7)

In [87]:
reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   review_id                99224 non-null  str  
 1   order_id                 99224 non-null  str  
 2   review_score             99224 non-null  int64
 3   review_comment_title     11568 non-null  str  
 4   review_comment_message   40977 non-null  str  
 5   review_creation_date     99224 non-null  str  
 6   review_answer_timestamp  99224 non-null  str  
dtypes: int64(1), str(6)
memory usage: 5.3 MB


In [88]:
reviews.describe()

,review_score
count,99224.000000
mean,4.086421
std,1.347579
min,1.000000
25%,4.000000
50%,5.000000
75%,5.000000
max,5.000000


In [89]:
reviews.describe(include="str")

,review_id,order_id,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
count,99224,99224,11568,40977,99224,99224
unique,98410,98673,4527,36159,636,98248
top,c444278834184f72b1484dfe47de7f97,c88b1d1b157a9999ce368f218a407141,Recomendo,Muito bom,2017-12-19 00:00:00,2017-06-15 23:21:05
freq,3,3,423,230,463,4


In [90]:
reviews.isnull().sum()

review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

In [91]:
reviews.duplicated().sum()

np.int64(0)

In [92]:
reviews.nunique()

review_id                  98410
order_id                   98673
review_score                   5
review_comment_title        4527
review_comment_message     36159
review_creation_date         636
review_answer_timestamp    98248
dtype: int64

In [93]:
reviews["review_score"].value_counts().sort_index()

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64

In [94]:
reviews.duplicated(
    subset=["review_id", "order_id"]
).sum()


np.int64(0)

In [95]:
reviews["order_id"].is_unique

False

#### Key analysis
- The table contains 99,224 rows and 7 columns.
- One row represents one review record associated with an order.
- There are no exact duplicate rows.
- `review_id` is not fully unique, and `order_id` is also not unique.
- The combination of `review_id` and `order_id` is unique, so it is the likely composite key.
- There are 98,673 unique orders represented in the reviews table.
- Review scores range from 1 to 5, with an average of about 4.09.
- Score 5 is the most common rating, followed by score 4.
- `review_comment_title` has 87,656 missing values and `review_comment_message` has 58,247 missing values.
- Missing review comments are likely valid because customers can submit a score without writing text.
- `review_creation_date` and `review_answer_timestamp` are currently stored as text and should later be converted to datetime.

# Sellers dataset

In [96]:
sellers = datasets["sellers"]
sellers.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [97]:
sellers.shape

(3095, 4)

In [98]:
sellers.info()

<class 'pandas.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype
---  ------                  --------------  -----
 0   seller_id               3095 non-null   str  
 1   seller_zip_code_prefix  3095 non-null   int64
 2   seller_city             3095 non-null   str  
 3   seller_state            3095 non-null   str  
dtypes: int64(1), str(3)
memory usage: 96.8 KB


In [99]:
sellers.describe()

,seller_zip_code_prefix
count,3095.000000
mean,32291.059451
std,32713.453830
min,1001.000000
25%,7093.500000
50%,14940.000000
75%,64552.500000
max,99730.000000


In [100]:
sellers.describe(include="str")

,seller_id,seller_city,seller_state
count,3095,3095,3095
unique,3095,611,23
top,3442f8959a84dea7ee197c632cb2df15,sao paulo,SP
freq,1,694,1849


In [101]:
sellers.isnull().sum()

seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

In [102]:
sellers.duplicated().sum()

np.int64(0)

In [103]:
sellers.nunique()

seller_id                 3095
seller_zip_code_prefix    2246
seller_city                611
seller_state                23
dtype: int64

In [104]:
sellers["seller_id"].is_unique

True

In [105]:
sellers["seller_state"].value_counts()

seller_state
SP    1849
PR     349
MG     244
SC     190
RJ     171
RS     129
GO      40
DF      30
ES      23
BA      19
CE      13
PE       9
PB       6
RN       5
MS       5
MT       4
RO       2
SE       2
AC       1
PI       1
MA       1
AM       1
PA       1
Name: count, dtype: int64

In [106]:
sellers["seller_city"].value_counts().head(10)

seller_city
sao paulo         694
curitiba          127
rio de janeiro     96
belo horizonte     68
ribeirao preto     52
guarulhos          50
ibitinga           49
santo andre        45
campinas           41
maringa            40
Name: count, dtype: int64

#### Key Analysis
- The table contains 3,095 rows and 4 columns.
- One row represents one seller registered on the platform.
- There are no missing values in any column.
- There are no duplicate rows.
- `seller_id` is unique and serves as the primary key.
- There are 611 unique seller cities and 23 seller states.
- Sellers are highly concentrated in São Paulo (SP), which accounts for 1,849 sellers (nearly 60% of all sellers).
- São Paulo city has the highest number of sellers (694), followed by Curitiba and Rio de Janeiro.
- Seller ZIP code prefixes have 2,246 unique values, indicating sellers are geographically distributed across many postal regions.

# Geolocation dataset

In [118]:
geolocation = datasets["geolocation"]

geolocation.head(2)

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP


In [108]:
geolocation.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  str    
 4   geolocation_state            1000163 non-null  str    
dtypes: float64(2), int64(1), str(2)
memory usage: 38.2 MB


In [109]:
geolocation.describe()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng
count,1.000163e+06,1.000163e+06,1.000163e+06
mean,3.657417e+04,-2.117615e+01,-4.639054e+01
std,3.054934e+04,5.715866e+00,4.269748e+00
min,1.001000e+03,-3.660537e+01,-1.014668e+02
25%,1.107500e+04,-2.360355e+01,-4.857317e+01
50%,2.653000e+04,-2.291938e+01,-4.663788e+01
75%,6.350400e+04,-1.997962e+01,-4.376771e+01
max,9.999000e+04,4.506593e+01,1.211054e+02


In [110]:
geolocation.describe(include="str")

,geolocation_city,geolocation_state
count,1000163,1000163
unique,8011,27
top,sao paulo,SP
freq,135800,404268


In [111]:
geolocation.isnull().sum()

geolocation_zip_code_prefix    0
geolocation_lat                0
geolocation_lng                0
geolocation_city               0
geolocation_state              0
dtype: int64

In [112]:
geolocation.duplicated().sum()

np.int64(261831)

In [113]:
geolocation.nunique()

geolocation_zip_code_prefix     19015
geolocation_lat                717360
geolocation_lng                717613
geolocation_city                 8011
geolocation_state                  27
dtype: int64

In [114]:
geolocation["geolocation_zip_code_prefix"].value_counts().head(10)

geolocation_zip_code_prefix
24220    1146
24230    1102
38400     965
35500     907
11680     879
22631     832
30140     810
11740     788
38408     773
28970     743
Name: count, dtype: int64

In [115]:
geolocation["geolocation_state"].value_counts()

geolocation_state
SP    404268
MG    126336
RJ    121169
RS     61851
PR     57859
SC     38328
BA     36045
GO     20139
ES     16748
PE     16432
DF     12986
MT     12031
CE     11674
PA     10853
MS     10431
MA      7853
PB      5538
RN      5041
PI      4549
AL      4183
TO      3576
SE      3563
RO      3478
AM      2432
AC      1301
AP       853
RR       646
Name: count, dtype: int64

In [116]:
geolocation["geolocation_city"].value_counts().head(10)

geolocation_city
sao paulo                135800
rio de janeiro            62151
belo horizonte            27805
são paulo                 24918
curitiba                  16593
porto alegre              13521
salvador                  11865
guarulhos                 11340
brasilia                  10470
sao bernardo do campo      8112
Name: count, dtype: int64

In [117]:
geolocation[["geolocation_lat", "geolocation_lng"]].describe()

,geolocation_lat,geolocation_lng
count,1.000163e+06,1.000163e+06
mean,-2.117615e+01,-4.639054e+01
std,5.715866e+00,4.269748e+00
min,-3.660537e+01,-1.014668e+02
25%,-2.360355e+01,-4.857317e+01
50%,-2.291938e+01,-4.663788e+01
75%,-1.997962e+01,-4.376771e+01
max,4.506593e+01,1.211054e+02


#### Key analysis
- The table contains 1,000,163 rows and 5 columns.
- One row represents one recorded latitude/longitude location for a ZIP-code prefix.
- There are no missing values.
- There are 261,831 exact duplicate rows, so duplicate removal should be considered during cleaning.
- `geolocation_zip_code_prefix` is not unique; there are 19,015 unique ZIP prefixes.
- Multiple coordinate records can exist for the same ZIP prefix, so the ZIP prefix cannot be used as a primary key in the raw table.
- There are 8,011 city-name values and 27 states.
- São Paulo has the largest number of geolocation records.
- City names are not fully standardized; for example, accented and unaccented versions such as `são paulo` and `sao paulo` appear separately.
- Latitude and longitude contain some extreme values outside Brazil’s normal geographic range, so coordinate outliers should be investigated during cleaning.
- For analytics, the geolocation table may later be aggregated to one representative latitude and longitude per ZIP-code prefix.

# Category Translation dataset

In [119]:
category_translation = datasets["category_translation"]

category_translation.head()

,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [120]:
category_translation.info()

<class 'pandas.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   product_category_name          71 non-null     str  
 1   product_category_name_english  71 non-null     str  
dtypes: str(2)
memory usage: 1.2 KB


In [121]:
category_translation.describe(include="str")

,product_category_name,product_category_name_english
count,71,71
unique,71,71
top,beleza_saude,health_beauty
freq,1,1


In [122]:
category_translation.isnull().sum()

product_category_name            0
product_category_name_english    0
dtype: int64

In [123]:
category_translation.duplicated().sum()

np.int64(0)

In [124]:
category_translation.nunique()

product_category_name            71
product_category_name_english    71
dtype: int64

In [125]:
category_translation["product_category_name"].is_unique

True

In [126]:
category_translation["product_category_name_english"].is_unique

True

#### Key analysis
- The table contains 71 rows and 2 columns.
- Each row maps one Portuguese product category name to one English category name.
- There are no missing values.
- There are no duplicate rows.
- `product_category_name` is unique and can be used as the primary key.
- `product_category_name_english` is also unique in this dataset.
- This table will be joined with the products table using `product_category_name`.
- The English category names will be useful for SQL analysis and Power BI dashboards.